### PR TITLE
DAOS-8905 control: Continue lib/hardware integration

### PR DESCRIPTION
**Pull request by mjmac**
Moves the PCI types from the common package into hardware,
changes the internal workings of PCIAddress to use numeric
fields, adds more test coverage.

### CODE REVIEW

File: src/control/lib/hardware/pci.go [hunk 1]
```diff
@@ -70,25 +111,50 @@ func (pa *PCIAddress) String() string {
...
+	fs := pa.FieldStrings()
+	return fmt.Sprintf("%s:%s:%s.%s", fs["Domain"], fs["Bus"], fs["Device"], fs["Function"])
 }
 
-// Equals compares string representation of address.
-func (pa *PCIAddress) Equals(o *PCIAddress) bool {
-	if pa == nil || o == nil {
+// Equals compares two PCIAddress structs for equality.
+func (pa *PCIAddress) Equals(other *PCIAddress) bool {
+	if pa == nil || other == nil {
```
Comment (kjacque) [inline src/control/lib/hardware/pci.go:120]: If both are nil, shouldn't they be equal?

File: src/control/lib/hardware/pci_test.go [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+		},
+		"add nil device": {
+			bus:    mockPCIBus(),
+			op:     "add",
+			expErr: isNilErr,
+		},
+		"nil bus contains device": {
+			dev: mockPCIDevice("test"),
+			op:  "nil contains",
+		},
```
Comment (kjacque) [inline src/control/lib/hardware/pci_test.go:493]: Personally would prefer to break Contains tests out into their own test function.

**Review by kjacque — DISMISSED** [review]
Nothing I'd block on.

File: src/control/lib/hardware/pci.go (hunk 1 above)
Comment (mjmac) [reply src/control/lib/hardware/pci.go:120]: I guess this is a bit of a philosophical question... To me, that would be somewhat surprising behavior. You can't really compare nothing to nothing, and maybe it's splitting hairs, but I think that this kind of comparison function should only operate on the values of a structure, not whether or not it's nil. To put it another way, I think I'd prefer that code which wants to assert that two addresses are nil should do so explicitly.

File: src/control/lib/hardware/topology.go [hunk 3]
```diff
@@ -166,57 +87,23 @@ func (n *NUMANode) AddDevice(dev *PCIDevice) {
...
-	for _, dev := range devices {
-		n.AddDevice(dev)
-	}
-	return n
+	return nil
 }
 
 // AddCore adds a CPU core to the node.
-func (n *NUMANode) AddCore(core CPUCore) {
+func (n *NUMANode) AddCore(core CPUCore) error {
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/lib/hardware/topology.go:98]: Yeah I deliberated on this as well. The first pass didn't have a return value and I was fine with that because the methods were used for test mocks. Then they started being used for production code and I wasn't crazy about silently ignoring errors. This specific method is very unlikely to return an error except for obvious programmer mistakes, but the other methods are more complicated and it seemed best to have this one follow the same pattern.

Generally speaking I think I'd prefer to default to returning errors for this kind of stuff, at least while we're shaking out the issues with the new library. Otherwise I'm concerned that we're going to swallow errors and waste time tracking things down because some weird issue was silently hidden.

File: src/control/lib/hardware/pci_test.go (hunk 2 above)
Comment (mjmac) [reply src/control/lib/hardware/pci_test.go:493]: Yeah, I got lazy. I'll do that in my next pass.

File: src/control/lib/hardware/pci.go (hunk 1 above)
Comment (kjacque) [reply src/control/lib/hardware/pci.go:120]: I can roll with that, although it's a little counterintuitive to me. But it's a corner case I don't think we're likely to hit with this structure outside unit tests anyway, so not a huge deal.

File: src/control/lib/hardware/topology.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/lib/hardware/topology.go:98]: That sounds reasonable to me, thanks.

**Review by tanabarr — APPROVED** [review]

File: src/control/lib/hardware/pci.go [hunk 4]
```diff
@@ -366,3 +389,109 @@ func NewPCIAddressSet(addrs ...string) (*PCIAddressSet, error) {
...
+
+type (
+	// PCIDevice represents an individual hardware device.
+	PCIDevice struct {
+		Name      string     `json:"name"`
+		Type      DeviceType `json:"type"`
+		NUMANode  *NUMANode  `json:"-"`
+		Bus       *PCIBus    `json:"-"`
+		PCIAddr   PCIAddress `json:"pci_address"`
+		LinkSpeed float64    `json:"link_speed"`
```
Comment (tanabarr) [inline src/control/lib/hardware/pci.go:401]: why do we have a link speed but not width? I would have imagined they would be stored together

**Review by tanabarr — DISMISSED** [review]

File: src/control/lib/hardware/pci.go (hunk 4 above)
Comment (mjmac) [reply src/control/lib/hardware/pci.go:401]: The hwloc info does not provide a link width value as far as I can tell. I believe that we'll need to incorporate support for a pcilib provider in order to obtain that detail.

**Review by tanabarr — DISMISSED** [review]

File: src/control/lib/hardware/pci_test.go [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+				t.Fatal(err)
+			}
+			addr2, err := NewPCIAddress(tc.addrStr2)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if diff := cmp.Diff(tc.expEqual, addr1.Equals(addr2)); diff != "" {
+				t.Fatalf("unexpected result (-want, +got):\n%s\n", diff)
+			}
```
Comment (kjacque) [inline src/control/lib/hardware/pci_test.go:157]: cmp.Diff feels a bit like overkill comparing two bools

**Review by kjacque — APPROVED** [review]
LGTM. I think this will produce some minor conflict with #7686 though, with the movement of PCIAddress to the hardware lib. Not sure which will land first but the other will need to clean that up.

**Review by tanabarr — APPROVED** [review]

File: src/control/lib/hardware/pci_test.go (hunk 5 above)
Comment (mjmac) [reply src/control/lib/hardware/pci_test.go:157]: Yeah, probably. I prefer the output format though. ;)
